File: src/main/kotlin/com/huawei/cangjie1/psi/CjCallExpression.java
package com.huawei.cangjie1.psi;

import com.google.common.collect.Lists;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;



public class CjCallExpression extends CjExpressionImpl implements CjCallElement, CjReferenceExpression {
    public CjCallExpression(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, D data) {
        return visitor.visitCallExpression(this, data);
    }

    @Override
    @Nullable
    public CjExpression getCalleeExpression() {
        return findChildByClass(CjExpression.class);
    }







}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjCallExpression.kt
package com.linqingying.cangjie.psi

import com.google.common.collect.Lists
import com.linqingying.cangjie.CjNodeTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement


class CjCallExpression(node: ASTNode) : CjExpressionImpl(node), CjCallElement, CjReferenceExpression {
    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitCallExpression(this, data)
    }

    override fun getLambdaArguments(): List<CjLambdaArgument> {
        return findChildrenByType<CjLambdaArgument>(CjNodeTypes.LAMBDA_ARGUMENT)
    }

    override fun getTypeArguments(): List<CjTypeProjection> {

        return typeArgumentList?.arguments ?: emptyList()
    }


    override fun getTypeArgumentList(): CjTypeArgumentList? {
        return findChildByType<PsiElement>(CjNodeTypes.TYPE_ARGUMENT_LIST) as CjTypeArgumentList?
    }

    override fun getCalleeExpression(): CjExpression? {
        return findChildByClass(CjExpression::class.java)
    }

    override fun getValueArgumentList(): CjValueArgumentList? {
        return findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST) as CjValueArgumentList?

    }

    override fun getValueArguments(): List<CjValueArgument> {

        val valueArgumentsInParentheses =
            valueArgumentList?.arguments ?: emptyList<CjValueArgument>()
        val functionLiteralArguments: List<CjLambdaArgument> =
            lambdaArguments
        if (functionLiteralArguments.isEmpty()) {
            return valueArgumentsInParentheses
        }
        val allValueArguments: MutableList<CjValueArgument> =
            Lists.newArrayList<CjValueArgument>()
        allValueArguments.addAll(valueArgumentsInParentheses)
        allValueArguments.addAll(functionLiteralArguments)
        return allValueArguments
    }


    override fun toString(): String {
        return node.elementType.toString()
    }
}
